runtime/domain.c

```diff
@@ -1524,32 +1531,26 @@ static void caml_poll_gc_work(void)
     CAML_EV_END(EV_MAJOR);
   }
 
-  if (atomic_load_acq(
-          (atomic_uintnat*)&Caml_state->requested_external_interrupt)) {
+  if (atomic_load_acq(&Caml_state->requested_external_interrupt)) {
     caml_domain_external_interrupt_hook();
   }
 
+  caml_reset_young_limit(Caml_state);
 }
 
 /* FIXME: do not raise async exceptions */
 void caml_handle_gc_interrupt(void)
 {
-  atomic_uintnat* young_limit = domain_self->interruptor.interrupt_word;
   CAMLalloc_point_here;
 
   CAML_EV_BEGIN(EV_INTERRUPT_GC);
-  if (atomic_load_acq(young_limit) == INTERRUPT_MAGIC) {
-    /* interrupt */
+  if (caml_incoming_interrupts_queued()) {
     CAML_EV_BEGIN(EV_INTERRUPT_REMOTE);
-    while (atomic_load_acq(young_limit) == INTERRUPT_MAGIC) {
-      uintnat i = INTERRUPT_MAGIC;
-      atomic_compare_exchange_strong(
-          young_limit, &i, (uintnat)Caml_state->young_start);
-    }
     caml_handle_incoming_interrupts();
     CAML_EV_END(EV_INTERRUPT_REMOTE);
   }
 
+  caml_reset_young_limit(Caml_state);
```

Comment:
Reading the whole diff, it's subtle that there is a call to `caml_reset_young_limit` here and also one at the end of `caml_poll_gc_work` called just below. Naively I wondered if we could keep just the reset at the end.

Here is my understand of what is going on:
- It would be wrong to just remove the reset here, because it would make `caml_poll_gc_work` empty the minor heap, even in the case where the limit was triggered only due to a STW interrupt and no GC work is in fact necessary.
- But we could remove the reset if there was a call to `caml_reset_young_limit` at the end of `caml_handle_gc_interrupts`. From a distance it sounds like that would be clearer than the current code. @gadmm, why did you add the reset here instead?

Reply:
It is there because the next line can still run OCaml code (finalisers) at this point. It is removed entirely in a later commit.